chain-extensions/impls/dapps-staking/src/lib.rs

```diff
@@ -0,0 +1,476 @@
+                let call_result = pallet_dapps_staking::Pallet::<T>::unbond_and_unstake(
+                    RawOrigin::Signed(caller).into(),
+                    contract,
+                    value,
+                );
+                return match call_result {
+                    Err(e) => {
+                        let mapped_error = DSError::try_from(e.error)?;
+                        let res = Result::<(), DSError>::Err(mapped_error);
+                        env.write(&res.encode(), false, None).map_err(|_| {
+                            DispatchError::Other("[ChainExtension] DappsStakingExtension UnbondAndUnstake failed to write result")
+                        })?;
+                        Err(e.error)
+                    }
+                    _ => {
+                        let res = Result::<(), DSError>::Ok(());
+                        env.write(&res.encode(), false, None).map_err(|_| {
+                            DispatchError::Other("[ChainExtension] DappsStakingExtension UnbondAndUnstake failed to write result")
+                        })?;
+                        Ok(())
+                    }
+                };
+            }
+
+            DappsStakingFunc::WithdrawUnbonded => {
+                let caller = env.ext().caller().clone();
+
+                let base_weight = <T as pallet_dapps_staking::Config>::WeightInfo::bond_and_stake();
+                env.charge_weight(base_weight)?;
+
+                let call_result = pallet_dapps_staking::Pallet::<T>::withdraw_unbonded(
+                    RawOrigin::Signed(caller).into(),
+                );
+                return match call_result {
+                    Err(e) => {
+                        let mapped_error = DSError::try_from(e.error)?;
+                        let res = Result::<(), DSError>::Err(mapped_error);
+                        env.write(&res.encode(), false, None).map_err(|_| {
+                            DispatchError::Other("[ChainExtension] DappsStakingExtension WithdrawUnbonded failed to write result")
+                        })?;
+                        Err(e.error)
+                    }
+                    _ => {
+                        let res = Result::<(), DSError>::Ok(());
+                        env.write(&res.encode(), false, None).map_err(|_| {
+                            DispatchError::Other("[ChainExtension] DappsStakingExtension WithdrawUnbonded failed to write result")
+                        })?;
+                        Ok(())
+                    }
+                };
+            }
+
+            DappsStakingFunc::ClaimStaker => {
+                let contract_bytes: [u8; 32] = env.read_as()?;
+                let contract = Self::decode_smart_contract(contract_bytes)?;
+                let base_weight = T::WeightInfo::claim_staker_with_restake()
+                    .max(T::WeightInfo::claim_staker_without_restake());
+                env.charge_weight(base_weight)?;
```

Comment:
I believe most users are using automatic restake option and in that case the weight of the call is higher. So you should first use `T::WeightInfo::claim_staker_with_restake`.

After receiving the result from executing the extrinsic call (line 305), you should fetch from it the real used weight and call `adjust_weight` on `env`.